text/0000-wallet-update-process.md

```diff
@@ -0,0 +1,169 @@
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Wallet Update Procedures
+
+The following sections outlines the processes used in the creation and update of outputs and related transactions within the wallet.
+
+### Transaction and Output Creation
+
+#### Coinbase Output + Transaction Creation
+
+Coinbase outputs and their related transactions are created via the wallet listener as part of the mining process. The workflow is:
+
+ * When mining, create a potential coinbase output for the target block.
+ * If the block is accepted and the output is detected in the UTXO set (via the [Update by Output process](#update-by-output-process), create a transaction log entry of type 'Confirmed Coinbase', with 'confirmed' set to true.
+
+#### Transacting (Payer -> Payee) Output + Transaction Creation
+
+Outputs and tranasctions are created during the transaction exchange process as follows. Note that the invoicing workflow is mostly identical with the roles of Payer and Payee reversed, so this workflow is not outlined separately.
+
+ * The sender creates a new 'blank' transaction Slate, adding inputs and change outputs to the slate.
+ * The sender generates a random kernel offset. (The kernel offset is always generated by the slate creator)
+ * The sender sends the slate to the payee (via file or http). If sending synchronously, (e.g. via http)  the associated transaction is saved to the log after a response from the payee's listener. If sending asynchronously, (e.g. via file), the transaction is saved immediately. 
+ * When saved, the associated transaction is set to type 'Sent Tx' with a status of 'unconfirmed', inputs are locked internally and a change output is added with status 'unconfirmed'.
+ * The payee receives the slate, creates (an) output(s) for the received amount with status unconfirmed, and immediately stores a transaction in their log of type 'Received Tx' with confirmed set to false. 
+ * The payee calculates and saves the transaction kernel commitment for later reference.
+ * The slate is returned to the payer for completion, who calculates and saves the transaction kernel. The transaction is sent to a node for validation.
+
+### Transaction and Output Statue Update Processes
+
+#### Update by Output Process
+
+The Update by Output Process is the main method by which the wallet updates its internal state, decides whether a transaction has been confirmed and when to remove or confirm individual outputs. This process works as follows:
+
+* For every unspent output in the wallet:
+* If the output's status in the user's wallet is 'unconfirmed' and it is present in the UTXO set, change the output's status to 'confirmed', and update the associated transaction status to 'confirmed'.
+* If the output's status in the user's wallet is 'confirmed' and it no longer appears in the UTXO set, set its status to 'Spent'. Note these outputs will usually be locked so they cannot be selected for spending again.
```

Comment:
Do you think it would be a good idea to create a new tx log entry when 'confirmed' outputs no longer exist in the UTXO set? This would give a more consistent feel for people who are using their seed on multiple wallets at the same time, as spending on wallet A would also appear as a spend on wallet B. Obviously this information would not be perfect but it would prevent situations where funds suddenly disappear without a trace.

Reply:
I think that could be a good enhancement. Right now the updater simply ignores the case where an output is not locked/doesn't have an associated transaction, i.e. it only updates a transaction if there's one to find. Could end up with a lot of spam in the tx log though since there's no real way to group outputs in this case, you'd just have to create a new TX for each instance of this occurring.

Reply:
We could group all outputs (without an associated tx log entry) that are no longer unspent in the same update cycle in 1 tx log entry. There is a chance that we group more than 1 tx but I don't think that is a big deal. It would however prevent a lot of spam.